### ISSUE TITLE
Crashes involving label color setters

### ISSUE THREAD

**Issue by pushfoo** [pyglet/pyglet#980]
### tl;dr

The following crash reliably on my system:

1. Setting any label's color to an RGB value via the setter
2. Setting an `HTMLLabel`'s color to an RGBA value the setter when it contains `<u>`, but not `<b>`

Both `master` and `development` are affected across all Python versions tried.

### Steps to reproduce

1.  Save the code below locally. It's also available via:
     * [this gist](https://gist.github.com/pushfoo/c313fcefdcdccbeda38d3614d5f75f72) 
     * wget + this [raw link](https://gist.github.com/pushfoo/c313fcefdcdccbeda38d3614d5f75f72/raw/192de503bc4f72d92b16265fe2f53795a0c16b91/pyglet_2_0_color_setter_crash.py)
3. Perform one or both of the following modifications:
    * Set `TAG='u'` in the code below
    * Set `COLOR=(255, 255, 255)` in the code below
4. Run it.

```Python
import pyglet

COLOR = (255, 255, 255, 255) # RGBA
LINE_HEIGHT_PX=40
FONT_SIZE_PT=20
TEXT="Test"
TAG='b' # u sets it off, but b does not

def html(tag, contents):
    return f"<{tag}>{contents}</{tag}>"

window = pyglet.window.Window(200,200)
batch = pyglet.graphics.Batch()

@window.event
def on_draw():
    window.clear()
    batch.draw()

shared_data = dict(x=window.height // 2, batch=batch)


label_data = [
    (pyglet.text.Label,     [TEXT]           , dict(bold=True, font_size=FONT_SIZE_PT)),
    (pyglet.text.HTMLLabel, [html(TAG, TEXT)], {}                                     )
]

label_instances = []

current_y = (window.height // 2) - (len(label_data) // 2) * 20
for label_class, args, kwargs in label_data:
    label = label_class(*args, y=current_y, **shared_data, **kwargs)
    label.color = COLOR
    label_instances.append(label)
    current_y += LINE_HEIGHT_PX

pyglet.app.run()
```

### Traceback

```console
Traceback (most recent call last):
  File "/home/user/src/pyglet/.scratch/brokencolorsetter.py", line 30, in <module>
    label.color = COLOR
  File "/home/user/src/pyglet/pyglet/text/__init__.py", line 274, in color
    self.document.set_style(0, len(self.document.text), {'color': color})
  File "/home/user/src/pyglet/pyglet/text/document.py", line 561, in set_style
    return super().set_style(0, len(self.text), attributes)
  File "/home/user/src/pyglet/pyglet/text/document.py", line 470, in set_style
    self.dispatch_event('on_style_text', start, end, attributes)
  File "/home/user/src/pyglet/pyglet/event.py", line 380, in dispatch_event
    if handler(*args):
  File "/home/user/src/pyglet/pyglet/text/layout.py", line 1482, in on_style_text
    self._update_color()
  File "/home/user/src/pyglet/pyglet/text/layout.py", line 1384, in _update_color
    _vertex_list.colors = colors[start:start + len(_vertex_list.colors)]
  File "/home/user/src/pyglet/pyglet/graphics/vertexdomain.py", line 75, in _attribute_setter
    attribute.buffer.set_region(self.start, self.count, data)
  File "/home/user/src/pyglet/pyglet/graphics/vertexbuffer.py", line 249, in set_region
    self._array[array_start:array_end] = data
ValueError: Can only assign sequence of same size
```

### System Information

| Detail | Value |
|---------|---------|
| OS     | Debian 11 / x86_64 |
| GPU  | Intel UHD Graphics 620 |
| DE     | Gnome 3 on Wayland |

I tried multiple Python versions. All seem to be affected. Here's their `python -m pyglet.info` output:

* [Python 3.9](https://gist.github.com/pushfoo/f492466a9e397460c2514d1d63338920)
* [Python 3.10](https://gist.github.com/pushfoo/7d2f8d7bd0224a2bf0faa935e8dd5433)
* [Python 3.12](https://gist.github.com/pushfoo/432dc7d985d1854dd660d90379a8666d)

**Comment by benmoran56** [comment]
Thanks for opening this.  This is actually a bug in the layout._update_color method. 

Recent changes to the Vertex Attribute setters have made them more strict. Before, it would silently allow you to set too little or too much data. It now crashes unless the passed data size is exact (as the ValueError shows). 

I believe we just need to append an extra 255 in the case or RGB color. I'll have a look at this shortly.

**Comment by benmoran56** [comment]
OK, so maybe it's not in that method in particular. We may need to handle it further up the stack, since <u> and <b> are handled differently.

**Comment by pushfoo** [comment]
> Recent changes to the Vertex Attribute setters have made them more strict.

I still haven't had the time to fully understand all the GL internals, but I think issues related to this were what I was trying to chase down with some of the proposed testing tools I mentioned earlier.

> We may need to handle it further up the stack, since `<u>` and `<b>` are handled differently.

I guessed this would be the case given the caret appeared to be a bare pair of vertices. Would you be up for allowing both the underline object and the caret to be regular `ShapeABC`s or other "batchable" / "drawable" items?

**Comment by benmoran56** [comment]
I made a quick fix to the Label setter so that it doesn't crash. `document.set_style` is still somewhat broken. This needs to be reviewed later, because the document style iterators and vertex lists don't appear to match up. Rather than passing silently, it will raise an exception now. 

The caret could be extended to make the vertices adjustable, but it would need some thought. It's best left to another discussion.
It's a standalone class, so it would be best to make a mock up for a first pass. Perhaps something to add to the `experimental` module.

**Comment by pushfoo** [comment]
1f14a3ec1 works as a temp fix for setting RGB on my side. :+1:

**Comment by benmoran56** [comment]
Great, i'll close this then.
